[RFC003] Nix-nickel (Draft, WIP) (#693)
tweag/nickel: rfcs/003-nix-nickel.md

- _Do not require unreasonable changes to Nix itself_. While it's probably
    illusory to avoid any change in Nix at all, we must strive to keep them
    reasonably small and undisturbing for regular Nix users, at least in the
    medium term.
- _Implementable in a reasonable timeline, if possible incrementally_. This RFC
    aims at proposing a solid vision for the future and not a temporary
    work-around. Still, given the finite bandwidth of the Nickel team, the
    proposed solution should be implementable in limited time (in the order of
    magnitude months, up to half a year). If possible, the plan should have
    intermediate milestones enabling new capabilities in an incremental way.
- _Backward-compatible_. We will realistically need to be able to leverage
    existing Nix code (especially from Nixpkgs) in some way: otherwise, even in
    the unlikely best case scenario where everybody suddenly switches to Nickel
    right away, it'll still take up a very long time to migrate everything.
- _Do not lock ourselves in the current Nixpkgs architecture_. Nixpkgs had
    made a number of design choices that are, in hindsight, not optimal (stdenv,

>     describing packages as functions instead of data, etc.). While we have to

Reviewer: https://github.com/nixpkgs-architecture/issues#1

### Scope

Nix expressions are used in related but different ways, each coming with varying
goals and constraints:

- **Nixpkgs**: derivations and packages in the original style.
- **Flakes**: new format for decentralized and composable packages.
- **NixOS modules**: NixOS system configuration.

In the long term, we aims at handling all those cases, but the scope of such an
undertaking appears very large for a single RFC. We decide to focus on the first
item: writing derivations and Nixpkgs-style packages.

The rationale is that derivations are the foundational concept on which the
whole Nix model is based. Flakes and NixOS modules build on derivations, as
illustrated by the following excerpt from a NixOS module:

```nix
systemd.myUnit.UnitConfig.ExecStart = pkgs.writeScript "foo" "..."
```

Something as innocent as using `pkgs.writeScript` in an hypothetical NixOS
Nickel module already implies to know how to generate derivations.

Thus, tackling derivations first makes sense. From there, Nixpkgs is probably
the thinnest layer over derivations of the three, and already open a lot of
possibilities, so we choose to include it as well.

## Challenges

### Interaction with Nixpkgs

> There a bunch of actions that require leveraging Nixpkgs:

Author: ```suggestion
There is a bunch of scenarios that require leveraging Nixpkgs:
```

for a package may be better separate and expose pure data and represent
computations as data dependencies.

Concretely, a derivation would be better represented simply as a recursive
records:

```nickel
builders.derivation = {

  # Interface

  name
    | doc "Name of the derivation, used in the Nix store path."
    | Str,

  version
    | doc "Version of the derivation, used in the Nix store path."
    | Str,
    | default = "",

  builder
    | doc "Command to be executed to build the derivation."
    | Path,

  args
    | doc "Arguments passed to the builder."
    | Array Str
    | default = [],

  outputs
    | doc "Symbolic names of the outputs of this derivation."
    | Array Str
    | default = ["out"],

  env
    | doc "Structured values passed to the builder."
    | {_: Str}
    # inherit doesn't exist in Nickel, but let's pretend, for conciseness
    = {inherit outputs},

  # Implementation

  drv
    | doc "The resulting store derivation."
    =
    (env & {
      name = "%{name}-%{version}";
      inherit builder args;
    })
    |> builtin.derivation
}

# ... other definitions building on this one

# actual package

> pkgs.hello = builders.unix_package & {

Author: True, and flakes are one example I think, where `output` is still a function but that's fine because the other metadata are one level up and therefore accessible without having to evaluate the output function. And back to our example, evaluating a merge and a simple function (which is morally a parametrized record) shouldn't make a big difference in practice.

One reason I could see to do that would be for consistency, as using both functions and merging may end up being confusing to the user, in particular which one is to be used in which case.

the Nix language of the future][nix-lang].

#### Package as data

Despite Nix being branded as a _functional_ package manager, a perhaps
surprising conclusion of Eelco Dolstra's document is that writing packages as
actual functions is retrospectively of questionable value. Functions need to be
applied (and thus arguments be produced) before we can access any data (which
hurts discoverability), their inputs are hard to override, etc. Overall,
functions are opaque _computations_, which makes them hard to inspect and to
patch. The reflection above pushes to switch to a model where packages are
rather _data_. Of course, computations still take place -- this is after all the
whole point of having a configuration language -- but the right representation
for a package may be better separated, exposing pure data and representing
computations as data dependencies.

Concretely, a derivation would be represented simply as a recursive record:

```nickel
builders.derivation = {

  # Interface

  name
    | doc "Name of the derivation, used in the Nix store path."
    | Str,

  version
    | doc "Version of the derivation, used in the Nix store path."

>     | Str,

Author: ```suggestion
    | Str
```

>     | default = "",

Reviewer 2: comma

    | Path,

  args
    | doc "Arguments passed to the builder."
    | Array Str
    | default = [],

  outputs
    | doc "Symbolic names of the outputs of this derivation."
    | Array Str
    | default = ["out"],

  env
    | doc "Structured values passed to the builder."
    | {_: Str}
    # inherit doesn't exist in Nickel, but let's pretend, for conciseness
    = {inherit outputs},

  # Implementation

  drv
    | doc "The resulting store derivation."
    =
    (env & {
      name = "%{name}-%{version}";
      inherit builder args;
    })
    |> builtin.derivation
}

# ... other definitions building on this one

# actual package
pkgs.hello = builders.unix_package & {
  name = "hello",
  version = "1.12",
  description = "A program that produces a familiar, friendly greeting",
  license = licenses.gpl,

  enable_gui
    | doc "Enable GTK+ support."
    | Bool
    | default false,

  src = builders.fetchurl & { url = ..., sha256 = ... },

  buildInputs = if enable_gui then [ gtk ] else [],
}
```

Thanks to laziness, `nix` could extract fields like `name` or `version` directly
without having to provide inputs or to evaluate `drv`.

> We call this approach the **PARM** (**P**ackage **A**s **R**records

Reviewer 3: Tiny typo :) 
```suggestion
We call this approach the **PARM** (**P**ackage **A**s **R**ecords
```

> We wall call this approach the **PARM** (**P**ackage **A**s **R**records

Reviewer 2: wall -> will

> We wall call this approach the **PARM** (**P**ackage **A**s **R**records

Author: ```suggestion
We call this approach the **PARM** (**P**ackage **A**s **R**records
```

**M**odel) thereafter.

#### Specifying dependencies

The `pkgs.hello` example unveils a difficulty about specifying dependencies. It
would work in a model where all the packages are defined in one place and
populate the same huge record that will give Nickelpkgs. But that would be
totally impracticable and anti-modular.

Writing the `pkgs.hello` in its own file in Nickel today would result in the
`gtk` occurrence raising an `unbound variable` error, and for a good reason:
`gtk` is used but seemingly defined nowhere. In practice, `gtk` is expected to
appear in the final recursive record that would Nickelpkgs. But from a language
point of view, allowing references to yet-to-be-defined identifiers, that is
having dynamic scoping, comes with its share of drawbacks: non-locality making
code harder to understand, inability to detect unbound variables statically
(incurring late error reporting), inability to properly typecheck such code,
etc.

Let us review possible suctions:

##### Explicit list

One possibility is to specify explicitly all the
dependencies as record fields without definition:

```nickel
# file hello.ncl
{
  # dependencies

>   gtk | NickelPackage,

Reviewer 2: this could allow inversion of control, where a child package can modify a parent package

```nickel
# file hello.ncl
{
  # dependencies
  gtk | NickelPackage = {
    version = "2.3.4", # bottom-up pinning. nixpkgs: gtk_2_3_4
    enable_somefeature = true, # require feature from parent
    # nixpkgs: hello = callpackage ./hello.nix { gtk = gtk.override { enable_somefeature = true; }; };
  },
```

problem: pollution &rarr; move to scope

```nickel
# file hello.ncl
{
  pkgs = {
    gtk | NickelPackage,
  },
```

Author: Good point. Although this could also be solved by merge priorities, the Nickelpkg layer erasing the local definitions. But it does sound fragile. Another issue is that having everything living at the same level forces all the versions of gtk to be the same in the final fixpoint, preventing local overrides (we want `this` package to have `python=python27` but `that` package to have `python=python3` for example).

For those reasons, I think your first suggestion sounds like the best way forward. Also, it could avoid having to repeat dependencies as say `inputs` and `build_inputs`: we could kill two bird with one stone and declare both inputs for Nix and the "merging interface" for Nickel as in

```nickel
{
  inputs = {
    gtk,
  },

 build_inputs = {
    foocompiler,
  }

  # ... inputs.gtk ...
}
```

Reviewer 2: bikeshed:
inputs sounds too general, could be packages and options (enable_somefeature)
packages is too long
pkgs is short & sweet (but also used in nixpkgs)

or we identify "package scopes" by type

```
{
  p | Pkgs = {
    gtk,
  },
  rundeps | Pkgs = [
    apkg, bpkg, cpkg, dpkg,
  ],
  drv = {
    build = m%"
      cp %{p.gtk}/lib/libgtk.so $out/lib/
      echo PATH=%{lib.makePath rundeps} >$out/bin/wrapper.sh
    "%m,
  },
}
```

such short names would compensate for nickel's missing `with` feature

would already apply the contract Package.

<!-- In this design, it would be nice to avoid having to repeat those
     dependencies in several places (like in inputs). -->

##### Dynamic scoping

<!-- DRAFT. not sure this is a good idea -->

A second solution would be to implement a form of dynamic scoping, but with a
special syntax, such that it doesn't impact normal uses of variables. Could
either have a special `self` or `super` keyword that refers to the final version
of the fixpoint, or a special syntax such as identifiers starting a specific
character:

```nickel
{
  name = "hello",
  version = "1.12",
  description = "A program that produces a familiar, friendly greeting",
  license = licenses.gpl,

  enable_gui
    | doc "Enable GTK+ support."
    | Bool
    | default false,

  src = builders.fetchurl & { url = ..., sha256 = ... },

  buildInputs = if enable_gui then [ super.gtk ] else [],

}
```

**TODO**: other solutions? The pkg subfield, but seems like a lesser version of
the first proposal. Bazel-like includes? From experience, the semantics is
confusing, it's tied to the filesystem structure. And in the end it's still a

> strange form of dynamic scoping: if we are to do it, let go for a proper version.

Reviewer 4: I agree with the confusing-ness of bazel targets, **but** they have a very strong point for them in that because since they are at a meta-language level, they are only resolved at the very last moment, meaning that
1. The actual path of the dependency really stays fully opaque, which
    1. Is more robust. For example, there was a couple of occurrences of code manipulating the output path of a dependency in nixpkgs, and that broke with CA derivations so I had to change it − like https://github.com/NixOS/nixpkgs/pull/124227
    2. (potentially) makes the final UX nicer because you don’t have ugly store paths leaking everywhere whenever you want to do something
2. Evaluating a target doesn’t require evaluating its dependencies (meaning that the whole discussion about packages-as-data vs packages-as-functions doesn’t even have to exist)

We anyways couldn’t do something like what bazel does (because in the Nix model you do need to resolve the paths of your dependencies to generate the build script, which isn’t the case in bazel), but maybe an in-between solution could be interesting

the semantics). This would preserve the Nickel advances, and we could leverage
directly any non-builtin Nix function, including all the Nixpkgs library
functions.

The potential drawbacks of this approach:

- Performance: Nickel is not expected to be in par with Nix at this point, so
  the performance of Nixpkgs compiled to Nickel might be prohibitively bad at
  first. On a positive note, this would provide a great benchmark suite and
  motivation to guide future performance improvements.
- Some builtin may incur effects, which are not trivial to combine with the
  execution model (revertible thunks). That being said, those effects are in
  practice idempotent and commutative (creating paths in the store, creating
  derivation, etc.), which should be fine.

Compilation could be done on the fly, as the compilation process would be rather
straightforward, or have a Nixpkgs snapshot pre-compiled to Nickel, or a mix of
both.

This solution requires to reimplement Nix builtins in Nickel (the nickel-nix
compatibility layer) . This is an interesting milestone in itself, because even
without a Nix-to-Nickel compiler, the compatibility layer would already make

> writing derivations in pure Nickel possible.

Reviewer 5: If it's a lot of work to reimplement builtins, I was wondering whether it may be possible to leverage the builtins within Nix itself (i.e. the executable / C++ source code) by using Nickel as a library in Nix. I'm not familiar with the Nix implementation, so it may be cumbersome, as often extension functions are tied closely to the interpreter. However, with some refactoring, a shared interface may be extracted for use in Nix proper and Nix-Nickel.

Using Nickel as a library seems like a good path for Nix in the future, whatever happens. Once that's in place, it makes sense to slowly port Nix from C++ to Rust. I believe there is already some interest in that direction.

Author: Yes, that is a possible path forward, but it is in opposition with the constraint `_Do not require unreasonable changes to Nix itself_`. We could debate if those changes are unreasonable, but I think that for a first approach, the less changes to Nix, the better. Not only from a disturbance/community point of view, but also I fear that even once accepted, the Nix side of the changes would be very long to land given the current development process, and that would hold us back.

- TODO: what about dynamic imports?

> - TODO: what about using a Nickel pkgs from Nix?

Reviewer 4: Unless (until?) we can get to a point where the Nix evaluator is just Nickel with a Nix frontend (in which case that problem won’t exist anymore because Nix could just call Nickel code transparently), I think that could be a limiting factor for the adoption of Nickel − at least for open-source projects, but these are the one with the most visibility. So we’ll want to have a decent story around this too

Author: As mentioned during the meeting, it seems going from a data representation to a function representation is easier, so we can hope for a simple generic wrappers around a Nickel package. But that needs to be investigated further before proceeding.

Reviewer 5: Are we back to considering a functional representation for Nickel pkgs?

Author: I think what is being discussed here is having a wrapper of some sort to use a Nickel package from Nix code, but just as a legacy compatibility mechanism. Using Nickel packages from another Nickel package would use the data representation and leverage all the benefits. However, if for some reason you need to use a package defined in Nickel from a Nixpkgs-like Nix codebase, you could, thanks to (very schematically) `to_nixpkg : Package -> NixPkgLegacyFunction`.

#### Using Nixpkgs in the PARM

Using a package from Nixpkgs in the PARM with a Nix-to-Nickel compiler would be
straightforward, as we could evaluate anything to a derivation whenever needed.
One would need to use Nixpkgs functions and idioms to do e.g. overrides though,
but this seems pretty hard to avoid, as a systematic translation from the

> Nixpkgs model to the PARM doesn't seem trivial, if even doable.

Reviewer 4: I don’t think that’s a bad thing in the long term tbh. Having an interface that would be too transparent could cause the path of least resistance to lead to a weird spaghetti of Nix and Nickel code intertangled, which would probably be a very bad thing.

At least, having some annoying constraints at the boundaries (even if just because both worlds have different idioms) mean that people will care about these boundaries, and things will stay clearly separated

Terraform would need to replicate this mechanism in some way.

We propose to adopt a more generic mechanism for strings with context, which
behavior can be parametrized. Strings become (conceptually) pairs `(s, ctxt)`
where `ctxt` is an arbitrary data structure with additional structure:

```Rust
combine: Ctxt -> Ctxt -> Ctxt,
pure: Str -> Ctxt,
pure_exp: Expr -> Ctxt,
```

The above functions could either be provided by user code or directly as an
interpreter plug-in.

- Existing standard strings would be equivalent to having a trivial context
  `null` and the obvious corresponding trivial structure.
- Custom contexts would be introduced by specific string delimiters, such as
    `nix%" %{pkgs.hello}/bin/hello"%`
- Different kind of strings are incompatible, to avoid accidentally forgetting
  or losing contexts. Also, we don't know a priori how to convert one context to
  another.
- There would be a contract to distinguish the different kind of
  strings in order enforce the usage of e.g. Nix style contexts for writing
  Nickel for Nix. We want to avoid users loosing or missing context unknowingly.

#### G-exps

> Alternative: something like `g-exp`. No magic, no extension, but less ergonomic.

Reviewer 5: What are `g-exprs`? Sounds like something Scheme-like (ala s-expression / fexprs). Perhaps a reference and an example of how it is less ergonomic would be helpful.

Author: Those are the Guix equivalent of string contexts, somehow: [manual page](https://guix.gnu.org/manual/en/html_node/G_002dExpressions.html). At this point this paragraph is just draft notes, but it'll deserve a proper link indeed once finalized. At a quick glance, it's less ergonomic because:

- Use a special syntax to start a g-exp `#~` (gexp)
- Use a special syntax to interpolate other g-exp `#$` (ungexp)

As opposed to Nix where string contexts just work without having the user to even think about it or convert between strings and derivations. On the other hand, the explicitness of g-exp may actually be a good point: arguably, it makes things less magic. But it would still be one additional thing that Nix users don't need to do today. Although, I think what was written at the time is not true: something akin to g-expr in Nickel using standard syntax for strings, like interpolation `%{}`, would still need a bit of magic and an extension mechanism, probably.

Reviewer 5: As I mentioned in IRL, my thinking is that the ideal solution is to have overloaded and customisable interpolated strings. Some languages do this such as Scala and Haskell (via quasiquotes and TH). In Scala it requires a prefix. Haskell doesn't have builtin interpolation, so quasiquotation kind of _is_ the way :man_shrugging:. I experimented a little bit with the the support in Idris the other night. It seems doable. Though I seem to have hit a limitation, misunderstanding, or bug in Idris.

Reviewer 5: In Idris, for instance, you can overload the standard string interpolation syntax. It should be possible to make something like this work:

```idris
postInstall : PkgExp
postInstall = """
    substituteInPlace \{out}/bin/my-script \\
      --replace /usr/bin/python \{python}/bin/python \\
      --replace /usr/bin/perl "\{perl}/bin/perl -I \{libfoo}"
  """
```

I'm not fond of that backslash as it evokes the idea of a "lambda". This is simply how string interpolation is defined in Idris (strange as `$` was discarded because of it's special meaning).

In [this short experiment](https://gist.github.com/steshaw/b3916d42f49f5b4e4ecc2922513cd7ad#file-pkg-idr-L50-L57), `pkg3a` should be equivalent to the desugared `pkg3` but I ran into a snag.
